larsim/EventGenerator/PhotonGen_module.cc

```diff
@@ -0,0 +1,250 @@
+    }
+
+    if (fScan) {
+      fPx = pset.get<double>("Px");
+      fPy = pset.get<double>("Py");
+      fPz = pset.get<double>("Pz");
+      std::cout << "Will generate photons from 3 points." << std::endl;
+    }
+  }
+
+  //____________________________________________________________________________
+  void PhotonGen::beginRun(art::Run& run)
+  {
+    art::ServiceHandle<geo::Geometry const> geo;
+    std::cout << "Number of optical detector: " << int(geo->Cryostat(0).NOpDet()) << std::endl;
+
+    //Boundary of Cryogenic
+    double CryoBounds[6];
+    geo->CryostatBoundaries(CryoBounds);
+    fXmin = CryoBounds[0];
+    fXmax = CryoBounds[1];
+    fYmin = CryoBounds[2];
+    fYmax = CryoBounds[3];
+    fZmin = CryoBounds[4];
+    fZmax = CryoBounds[5];
+    std::cout << "Cryo Boundaries:" << std::endl;
+    std::cout << "Xmin: " << fXmin << " Xmax: " << fXmax << " Ymin: " << fYmin << " Ymax: " << fYmax
+              << " Zmin: " << fZmin << " Zmax: " << fZmax << std::endl;
+    run.put(std::make_unique<sumdata::RunData>(geo->DetectorName()));
+
+    return;
+  }
+
+  //----------------------------------------------------------------
+  void PhotonGen::produce(art::Event& evt)
+  {
+    std::mt19937 rng;
+    rng.seed(std::random_device()());
+    std::uniform_real_distribution<double> distX(fXmin, fXmax);
+    std::uniform_real_distribution<double> distY(fYmin, fYmax);
+    std::uniform_real_distribution<double> distZ(fZmin, fZmax);
+    std::uniform_real_distribution<double> width(-2.0, 2.0);
+    srand((unsigned)time(NULL));
+
+    std::unique_ptr<std::vector<simb::MCTruth>> truthcol(new std::vector<simb::MCTruth>);
+    simb::MCTruth truth;
+    truth.SetOrigin(simb::kSingleParticle);
+
+    if (fScan) {
+      fX = distX(rng);
+      fY = fPy + width(rng);
+      fZ = fPz + width(rng);
+      Sample(truth);
+    }
+    else {
+      fX = distX(rng);
+      fY = distY(rng);
+      fZ = distZ(rng);
+
+      Sample(truth);
+    }
+
+    truthcol->push_back(truth);
+    evt.put(std::move(truthcol));
+    return;
+  }
+
+  void PhotonGen::Sample(simb::MCTruth& mct)
+  {
+    std::cout << "Photons Shooting at the Position: " << fX << " " << fY << " " << fZ << std::endl;
+
+    CLHEP::RandFlat flat(fEngine);
+    CLHEP::RandGaussQ gauss(fEngine);
+
+    for (int j = 0; j != fN; ++j) {
+      TVector3 pos;
+      pos[0] = fX;
+      pos[1] = fY;
+      pos[2] = fZ;
+
+      double time;
+      if (fTDist == kGAUS) { time = gauss.fire(fT, fSigmaT); }
+      else {
+        time = fT + fSigmaT * (2.0 * flat.fire() - 1.0);
+      }
+      fShotPos = TLorentzVector(pos[0], pos[1], pos[2], time);
+
+      //momentum (supplied in eV, convert to GeV)
+      double p = fP;
+      if (fPDist == kGAUS) { p = gauss.fire(fP, fSigmaP); }
+      else {
+        p = fP + fSigmaP * (2.0 * flat.fire() - 1.0);
+      }
+      p /= 1000000000.;
+
+      //angles
+      double costh = 2 * flat.fire() - 1;
+      double sinth = pow(1 - pow(costh, 2), 0.5);
```

Comment:
The following will be more efficient:
```suggestion
      double sinth = std::sqrt(1 - costh * costh);
```